weld/subsystem/src/main/java/org/jboss/as/weld/WeldBootstrapService.java

```diff
@@ -236,8 +305,91 @@ public boolean isStarted() {
+        //we need to add the listener first, but the listener may be invoked before the CAS to never
+        CompletableFuture<Boolean> attemptingBounce = new CompletableFuture();
+
+        try {
+            CompletableFuture<Boolean> listenerDone = new CompletableFuture<>();
+            LifecycleListener listener = new LifecycleListener() {
+                @Override
+                public void handleEvent(final ServiceController<?> controller, final LifecycleEvent event) {
+                    try {
+                        try {
+                            if (controller.getServiceContainer().isShutdown() || !attemptingBounce.get()) {
+                                controller.removeListener(this);
+                                return;
+                            }
+                        } catch (InterruptedException | ExecutionException e) {
+                            throw new RuntimeException(e);
+                        }
+                        if (deploymentController.getMode() != ServiceController.Mode.ACTIVE ||
+                                controller.getMode() == ServiceController.Mode.REMOVE) {
+                            return;
+                        }
+                        if (event == LifecycleEvent.DOWN) {
+                            controller.removeListener(this);
+                            do {
+                                if (controller.getMode() != ServiceController.Mode.NEVER) {
+                                    return;
+                                }
+                            } while (!controller.compareAndSetMode(ServiceController.Mode.NEVER, ServiceController.Mode.ACTIVE));
+                        }
+                    } finally {
+                        listenerDone.complete(true);
+                    }
+                }
+            };
+            //
+            controller.getServiceContainer().addService(controller.getName().append("fakeStabilityService")).setInstance(new Service() {
+                @Override
+                public void start(StartContext context) throws StartException {
+                    context.asynchronous();
+                    listenerDone.handle(new BiFunction<Boolean, Throwable, Object>() {
+                        @Override
+                        public Object apply(Boolean aBoolean, Throwable throwable) {
+                            context.getController().setMode(ServiceController.Mode.REMOVE);
+                            context.complete();
+                            return null;
+                        }
+                    });
+                }
+
+                @Override
+                public void stop(StopContext context) {
+
+                }
+            }).install();
+            controller.addListener(listener);
+            if (!controller.compareAndSetMode(ServiceController.Mode.ACTIVE, ServiceController.Mode.NEVER)) {
+                controller.removeListener(listener);
```

Comment:
@stuartwdouglas Can this result in listenerDone not completing and fakeStabilityService leaking?

Reply:
Yes, fixed. 

This whole thing is a horrible hack.